### PR TITLE
Group17

### PR DESCRIPTION
**Pull request by sitarakb**
ONLY LOOK AT GAME.CPP & GAME.HPP (other commits are from the branch this was made from).

Other contributions: Game.cpp & Game.hpp that's currently in main (from the statsMenu expedited PR)

### CODE REVIEW

**Comment by codeinputmachine** [comment]
Hey @sitarakb, Code Input detected this PR has a merge conflict. Several conflicts of this PR can be resolved with a semantic merge driver. Code Input can do that automatically: https://codeinput.com/r/BZZsXgk9I0K Let me know if you need more help with this conflict or how Code Input works.

**Comment by marksandychez** [comment]
> Hey @sitarakb, Code Input detected this PR has a merge conflict. Several conflicts of this PR can be resolved with a semantic merge driver. Code Input can do that automatically: https://codeinput.com/r/BZZsXgk9I0K Let me know if you need more help with this conflict or how Code Input works.

fixed these.

**Comment by karensuzue** [comment]
**Reviewers for Group 17:** Please select all commits starting from **[5be3ebe](https://github.com/CSE498/Spring2026-CompanyA/pull/90/commits/5be3ebecaf817645fe6de1b992bcd529b78e1b52) and earlier**

File: source/Interfaces/GUI/interface/Game.cpp [hunk 1]
```diff
@@ -10,230 +10,339 @@
 #include <iostream>
 #include "../../../core/AgentBase.hpp"
 
-namespace cse498 {
+namespace cse498
+{
 
-class StubAgent : public AgentBase {
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:15]: This temporary stub agent probably should not live inside Game.cpp; it adds extra test/demo code to a production game file.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 2]
```diff
@@ -10,230 +10,339 @@
...
-public:
-    StubAgent(size_t id, const std::string& name, const WorldBase& world) : AgentBase(id, name, world) {
-        SetSymbol('@'); // represents player on screen
-    }
+    constexpr int TURN_DELAY = 100;
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:22]: constexpr is fine here, but this constant is only useful for the stub class and does not seem tied to the actual game loop logic.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 3]
```diff
@@ -10,230 +10,339 @@
...
+            auto result = mImageManager->LoadImage(name, path);
+            if (!result)
+            {
+                std::cerr << "LoadImage failed: " << result.error() << std::endl;
+                return false;
+            }
+            return true;
+        };
 
-bool Game::Initialize() {
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:34]: Asset loading is very repetitive here; this would be easier to maintain if the paths were grouped into tables/arrays and loaded in a loop.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 4]
```diff
@@ -10,230 +10,339 @@
...
-    if (!mGameView->Initialize())
-        return false;
+        // Grass variants
+        if (!LoadCheck("grass", std::string(ASSETS_DIR) + "/" + + "world/forest/floor_tiles/tile_grass_1.png"))
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp]: The repeated + "/" + + "..." path building is messy and error-prone; it should be cleaned up before merge.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 5]
```diff
@@ -10,230 +10,339 @@
...
+
+        // Item pickup notifications
+        mPickupText.SetRenderer(renderer);
+        mPickupText.SetSize(20);
+
+        // Stats text
+        mStatsText.SetRenderer(renderer);
+
+        // Set up image manager and load all tile assets
+        mImageManager = std::make_unique<ImageManager>(renderer);
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:58]: Good use of std::make_unique here since ownership is clear and it avoids raw new.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 6]
```diff
@@ -10,230 +10,339 @@
...
+
+        // Helper lambda to load and propagate errors
+        auto LoadCheck = [&](const std::string &name, const std::string &path) -> bool
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:61]: This lambda is a good use of modern C++ because it removes repeated error-handling code and keeps initialization cleaner.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 7]
```diff
@@ -10,230 +10,339 @@
...
 
-    // Helper lambda to load and propagate errors
-    auto LoadCheck = [&](const std::string& name, const std::string& path) -> bool {
-        auto result = mImageManager->LoadImage(name, path);
-        if (!result) {
-            std::cerr << "LoadImage failed: " << result.error() << std::endl;
-            return false;
-        }
+        // Analytics setup
+        mAnalyticsManager = std::make_shared<AnalyticsManager>();
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:197]: shared_ptr may be okay here, but it would be good to confirm that ownership is actually shared and not just using shared_ptr by default.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 8]
```diff
@@ -10,230 +10,339 @@
...
+    // -----------------------------------------------------------------------
 
-void Game::Run() {
-    mRunning = true;
-    while (mRunning && mState != GameState::QUIT) {
-        HandleEvents();
+    void Game::Run()
+    {
+        mRunning = true;
+        while (mRunning && mState != GameState::QUIT)
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:408]: The main loop structure is reasonable, but the duplicated code around it makes the control flow harder to trust.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 9]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
+                            int cols = static_cast<int>(Inventory::ITEMS_PER_ROW);
+                            switch (event.key.keysym.sym) {
+                            case SDLK_w: mBackpackCursorRow = (mBackpackCursorRow - 1 + rows) % rows; break;
+                            case SDLK_s: mBackpackCursorRow = (mBackpackCursorRow + 1) % rows; break;
+                            case SDLK_a: mBackpackCursorCol = (mBackpackCursorCol - 1 + cols) % cols; break;
+                            case SDLK_d: mBackpackCursorCol = (mBackpackCursorCol + 1) % cols; break;
+                            default: break;
+                            }
+                        } else {
+                            static Uint32 last_move_time = 0;
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:523]: Using a function-local static for movement cooldown works, but it hides state inside the handler and makes timing behavior harder to test.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 10]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
                 case SDLK_d:
-                    if (mState == GameState::OVERWORLD || mState == GameState::DUNGEON) {
-                        static Uint32 last_move_time = 0;
-                        Uint32 now = SDL_GetTicks();
-                        if (now - last_move_time >= 150) {
-                            ProcessPlayerMove(event.key.keysym.sym);
-                            last_move_time = now;
+                    if (mState == GameState::OVERWORLD || mState == GameState::DUNGEON)
+                    {
+                        if (mShowBackpack) {
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:511]: This input handler is taking on a lot of responsibilities at once: menu input, world movement, backpack navigation, slot swapping, and pause logic.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 11]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
+                    {
+                        // SDLK_1 = slot 0, SDLK_2 = slot 1, etc.
+                        size_t hotbar_slot = (event.key.keysym.sym == SDLK_0) ? 9 : static_cast<size_t>(event.key.keysym.sym - SDLK_1);
+                        if (hotbar_slot < Inventory::HOTBAR_SIZE) {
+                            size_t backpack_index = Inventory::HOTBAR_SIZE
+                                + mBackpackCursorRow * static_cast<int>(Inventory::ITEMS_PER_ROW)
+                                + mBackpackCursorCol;
+
+                            Inventory& inv = (mState == GameState::OVERWORLD)
+                                ? mOverworldPlayer->GetInventory()
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:562]: Nice that this avoids duplicated swap logic, but it still depends on a large conditional branch in an already crowded event handler.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 12]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
+    void Game::UpdateMainMenu() {}
 
-    // Max camera position so the viewport never scrolls past the grid edge
-    int max_cam_x = std::max(0, static_cast<int>(grid.GetWidth()) - Tiles_x);
-    int max_cam_y = std::max(0, static_cast<int>(grid.GetHeight()) - Tiles_y);
+    void Game::UpdateOverworld()
+    {
+        // skip the player in the world agent list, they should choose their own move when needed to.
+        if (mTurnTaken) {
+            for (size_t i = 0; i < mOverWorld->GetNumAgents(); ++i) {
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:627]: Skipping the player explicitly is reasonable, but it would be cleaner if the world/agent system exposed a clearer way to update only non-player agents.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 13]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
-void Game::RenderOverworld() {
-    // Layer 0 — Tiles (bottom)
-    RenderWorld(*mOverworldGrid, mCamX, mCamY);
+            int screen_x = (static_cast<int>(pos.CellX()) - mCamX) * tw;
+            int screen_y = (static_cast<int>(pos.CellY()) - mCamY) * th;
 
-    // Layer 1 — items/objects on the ground
-    // RenderItems();
+            // Pick sprite based on whether this is the player or an NPC
+            const std::string &sprite = (&agent == mOverworldPlayer) ? "player" : mOverWorld->GetAgentSpriteName();
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp]: Rendering works, but sprite selection logic inside the render loop couples drawing too tightly to game object identity.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 14]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
+            mImageManager->DrawImage(sprite, screen_x, screen_y, tw, th);
+        }
 
-    // Layer 2 — agents/NPCs
-    int tw = static_cast<int>(mOverworldGrid->GetTileWidth());
-    int th = static_cast<int>(mOverworldGrid->GetTileHeight());
+        // Layer 2 — UI/HUD
+        RenderHotbar(mOverworldPlayer->GetInventory());
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:717]: Good feature addition, but these UI rendering functions are pretty large and may deserve their own helper or UI component file later.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 15]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
+        mImageManager->DrawImage("inventory_bar", bar_x, bar_y, bar_w, bar_h);
+
+        // Draw items in each hotbar slot
+        const auto& slots = inventory.GetInventoryArray();
+        int slot_size = bar_w / static_cast<int>(Inventory::HOTBAR_SIZE);
+
+        for (size_t i = 0; i < Inventory::HOTBAR_SIZE; ++i) {
+            const auto& slot = slots[i];
+            if (!slot.IsEmpty()) {
+                const Item* item = slot.GetItem();
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:886]: Since this depends on raw pointers returned from inventory slots, it would be good to make sure ownership/lifetime is clearly guaranteed by the inventory system.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 16]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
-        int max_cam_x = std::max(0, static_cast<int>(mOverworldGrid->GetWidth()) - Tiles_x);
-        int max_cam_y = std::max(0, static_cast<int>(mOverworldGrid->GetHeight()) - Tiles_y);
+        mPickupText.SetContent(mPickupMessage);
+        int text_x = (w - mPickupText.GetWidth()) / 2;
+        mPickupText.Draw(text_x, 20);
+    }
 
-        mCamX = std::clamp(mPlayerX - Tiles_x / 2, 0, max_cam_x);
-        mCamY = std::clamp(mPlayerY - Tiles_y / 2, 0, max_cam_y);
+    size_t Game::KeyToAction(SDL_Keycode key) {
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:980]: This helper is a good cleanup choice because it centralizes input-to-action mapping instead of repeating it.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 17]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
-            case SDLK_a:
-                mDungeonPlayerX = std::max(0, mDungeonPlayerX - 1);
-                break;
-            case SDLK_d:
-                mDungeonPlayerX = std::min(max_x, mDungeonPlayerX + 1);
-                break;
-            default:
-                break;
+        if (mState == GameState::OVERWORLD) {
+            mOverWorld->DoAction(*mOverworldPlayer, action);
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:995]: This is cleaner than manually changing player coordinates, since it lets movement go through the world/action system instead of bypassing it.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 18]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
+
+            mCamX = std::clamp(mPlayerX - Tiles_x / 2, 0, max_cam_x);
+            mCamY = std::clamp(mPlayerY - Tiles_y / 2, 0, max_cam_y);
         }
 
-        int tw = static_cast<int>(mDungeonGrid->GetTileWidth());
-        int th = static_cast<int>(mDungeonGrid->GetTileHeight());
+        else if (mState == GameState::DUNGEON) {
+            // Snapshot inventory count before move
+            size_t items_before = 0;
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:1014]: The inventory change check works, but scanning the full inventory before and after every move may be more work than needed for a simple pickup notification.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 19]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
 
-        mDungeonCamX = std::clamp(mDungeonPlayerX - Tiles_x / 2, 0, max_cam_x);
-        mDungeonCamY = std::clamp(mDungeonPlayerY - Tiles_y / 2, 0, max_cam_y);
-    }
+            WorldPosition pos = mDungeonPlayer->GetLocation().AsWorldPosition();
+            mDungeonPlayerX = static_cast<int>(pos.CellX());
+            mDungeonPlayerY = static_cast<int>(pos.CellY());
+
+            // Detect level change — player was moved to (1,1) and grid was regenerated
+            if (pos.CellX() == 1 && pos.CellY() == 1 && pos_before.CellX() != 1 && pos_before.CellY() != 1) {
```
Comment (Luke-Antone) [inline source/Interfaces/GUI/interface/Game.cpp:1029]: This level-change check relies on hard-coded coordinates, which feels fragile and should probably be tied to a clearer game/world event.

**Review by Luke-Antone — COMMENTED** [review]
Advanced C++:
The file shows some solid use of modern C++ features, especially with std::make_unique, lambdas like LoadCheck, and helper functions such as KeyToAction that reduce repeated code. That said, the advanced features here feel fairly light overall, and in a few places the design should be checked more closely to make sure the choices match the actual ownership model, especially with std::shared_ptr for analytics. My main takeaway is that the file does use modern C++ in helpful ways, but the stronger focus should be on using those features consistently and only where they clearly improve safety, ownership, and maintainability.

Implementation Quality:
The biggest concern in this file is implementation quality. The diff appears to contain duplicated function bodies and merge-like leftovers in major functions such as Initialize, SetupOverworld, SetupDungeon, menu setup, rendering, and update logic, which makes the code hard to follow and lowers confidence in the final structure. Beyond that, Game.cpp is taking on a lot of responsibilities at once, including asset loading, world setup, input handling, movement, rendering, inventory UI, and stats, so the file feels too large and tightly coupled. There are good ideas here, but before merge this code would benefit from cleanup of duplicated sections, clearer separation of responsibilities, and reducing fragile hard-coded logic.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 20]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
 
-        int max_cam_x = std::max(0, static_cast<int>(mOverworldGrid->GetWidth()) - Tiles_x);
-        int max_cam_y = std::max(0, static_cast<int>(mOverworldGrid->GetHeight()) - Tiles_y);
+        mPickupText.SetContent(mPickupMessage);
+        int text_x = (w - mPickupText.GetWidth()) / 2;
+        mPickupText.Draw(text_x, 20);
+    }
 
-        mCamX = std::clamp(mPlayerX - Tiles_x / 2, 0, max_cam_x);
-        mCamY = std::clamp(mPlayerY - Tiles_y / 2, 0, max_cam_y);
```
Comment (LimEddie19) [inline source/Interfaces/GUI/interface/Game.cpp:652]: You could change Tiles_X and Tiles_Y to tiles_x and tiles_y for consistency

File: source/Interfaces/GUI/interface/Game.cpp [hunk 21]
```diff
@@ -10,230 +10,339 @@
...
-// -----------------------------------------------------------------------
+        std::cout << "Working directory: " << std::filesystem::current_path() << std::endl;
+        std::cout << "Asset Dir: " << std::string(ASSETS_DIR) << std::endl;
+
+        SDL_Renderer *renderer = mGameView->GetRenderer();
+
+        // Title text
+        mTitleText.SetRenderer(renderer);
+        mTitleText.SetContent("Slay the Dungeon");
+        mTitleText.SetSize(48);
```
Comment (LimEddie19) [inline source/Interfaces/GUI/interface/Game.cpp:41]: I've seen the 48 used for text size and icon size

File: source/Interfaces/GUI/interface/Game.cpp [hunk 22]
```diff
@@ -319,374 +434,545 @@ void Game::HandleEvents() {
...
-            case SDLK_a:
-                mPlayerX = std::max(0, mPlayerX - 1);
-                break;
-            case SDLK_d:
-                mPlayerX = std::min(max_x, mPlayerX + 1);
-                break;
-            default:
-                break;
+        Uint32 elapsed = SDL_GetTicks() - mPickupMessageTime;
+        if (elapsed > 1000) {
```
Comment (LimEddie19) [inline source/Interfaces/GUI/interface/Game.cpp:968]: Although the "1000" was not used extensively. It could be named for easier understanding.

File: source/Interfaces/GUI/interface/Game.cpp
Comment (LimEddie19) [inline source/Interfaces/GUI/interface/Game.cpp:1]: **Informative identifiers**:
All names are easily identifiable and make sense logically. A few names could be improved for consistency, like Tiles_x and Tiles_y, could probably be tiles_x and tiles_y.

**Avoids magic numbers**:
There are a few numbers like 64 which you explained in the comments as the "64 x 64 tile size". There are extensive use of this number and others like it throughout the file. I think turning it into a named constant could help readability. but I'm not sure if that constitutes as a magic number as some of them make a lot of sense from context. I will highlight the number that I think are possible magic numbers. Of course, feel free to correct me.

**Intuitive and functional interface**:
The class has clear high-level methods for setup, update, rendering, and state changes. 

**Informative comments**:
The comments are generally helpful, especially the section headers and explanations for rendering layers, asset groups, and movement logic.

**Consistent style**:
The style is consistent Function names and member variables using the mCamelCase. A few names could be improved for consistency, like Tiles_x and Tiles_y, could probably be tiles_x and tiles_y.

**Visual appearance**:
The code is visually easy to understand and well organized with section dividers. 

Overall, the code looks great. Just a minor few suggestions. There were many great uses of constexpr, lambdas, and value-semantics in your code.

File: source/Interfaces/GUI/interface/Game.hpp [hunk 23]
```diff
@@ -13,254 +13,289 @@
...
+        GameState mPreviousState = GameState::MAIN_MENU; /// Used to resume after pause
+
+        // -------------------------
+        // Constants
+        // -------------------------
+        static constexpr int kDefaultWindowWidth = 800;
+        static constexpr int kDefaultWindowHeight = 600;
+
+        static constexpr int kInitialPlayerX = 1;
+        static constexpr int kInitialPlayerY = 1;
```
Comment (LimEddie19) [inline source/Interfaces/GUI/interface/Game.hpp:74]: Great use of constexpr.

File: source/Interfaces/GUI/interface/Game.cpp (hunk 4 above)
Comment (LimEddie19) [inline source/Interfaces/GUI/interface/Game.cpp]: Adding onto Luke's suggestion, you could create a struct for the names and paths of the asset and store them in an array, then just loop over them so it looks cleaner.

File: source/Interfaces/GUI/interface/Game.cpp [hunk 24]
```diff
@@ -10,230 +10,339 @@
...
-}
+        mMainMenu.AddOption("New Game", [this]() { TransitionTo(GameState::OVERWORLD); });
 
-void Game::SetupPauseMenu() {
-    mPauseMenu.Clear();
+        mMainMenu.AddOption("Settings", [this]() { TransitionTo(GameState::SETTINGS); });
 
-    mPauseMenu.AddOption("Resume", [this]() { Resume(); });
+        mMainMenu.AddOption("Quit", [this]() { Quit(); });
+    }
```
Comment (LimEddie19) [inline source/Interfaces/GUI/interface/Game.cpp:372]: Great use of lambdas

File: source/Interfaces/GUI/interface/Game.cpp [hunk 25]
```diff
@@ -10,230 +10,339 @@
...
+        // Map every WorldGrid cell type name to its matching image name
+        for (size_t y = 0; y < world_h; ++y)
+        {
+            for (size_t x = 0; x < world_w; ++x)
+            {
+                WorldPosition pos(x, y);
+                const std::string &cell_name = grid.GetCellTypeName(grid[pos]);
+                mOverworldGrid->SetCell(x, y, cell_name);
+            }
         }
```
Comment (LimEddie19) [inline source/Interfaces/GUI/interface/Game.cpp:288]: This could be made a helper function since lines R266 to R274 uses the same exact logic to build the grid

Comment (AbbyFranz) [reply source/Interfaces/GUI/interface/Game.cpp]: R286 to R293 also uses this same logic

File: source/Interfaces/GUI/interface/Game.cpp
Comment (AbbyFranz) [inline source/Interfaces/GUI/interface/Game.cpp:309]: This discord discussion is out of date. 

Due to issues with the player character, the player is now created directly within the dungeonworld class, and any other enemy agents will be created there as well. 

-I'm not sure if this has been addressed since I realize this is an old commit, but feel free to msg group15 with any questions about this.

Comment (AbbyFranz) [inline source/Interfaces/GUI/interface/Game.cpp:624]: UpdateOverworld and UpdateDungeon are essentially duplicate code. They could be made into one function where you just indicate which world you are updating on use.

Comment (AbbyFranz) [inline source/Interfaces/GUI/interface/Game.cpp:980]: This function can be made const

Comment (AbbyFranz) [inline source/Interfaces/GUI/interface/Game.cpp:1009]: This and R964 to R972 is duplicate code. You could probably make an update camera helper

Comment (AbbyFranz) [inline source/Interfaces/GUI/interface/Game.cpp:1039]: "Uses standard library algorithms" - According to rubric/guidelines you could use std::accumulate and a lambda here, but I also think that might be overkill here and end up hurting readabilty.

Comment (AbbyFranz) [inline source/Interfaces/GUI/interface/Game.cpp:31]: Since returning false indicates an initialization error, each path that returns false should print out a cerr message like you have on R60. Other paths returninf false such as this one and many of load checks are missing an error checking message.

Comment (AbbyFranz) [inline source/Interfaces/GUI/interface/Game.cpp]: Hardcoding - This uses "dun_monster" for the image file if the actor is not the player, however this doesn't account for there being different types of monsters with different sprites within the game

**Review by AbbyFranz — COMMENTED** [review]
Implementation
1. No warnings during compilation.- N/A - There were so many errors during compiliation from many files across the company this it was impossible to tell where they came from. Ignoring warnings, the demo compiled and ran on my machine with no issues :)
2. Proper uses of namespaces - Good!
3. Adheres to the C++ standard - Good!
4. Appropriate Modularity. - Theres quite a bit of code duplication. Specifically, there singular functions that do the same thing to overworld vs dungeonworld, where alternatively it could be one function that takes an argument to determine which world to work with
5. Uses modern C++ constructs - Good!
6. Uses standard library algorithms. -Good! there was one spot that I indicated it could be used, but it is probably unnecessary
7. Pointer handling. - Good!
8. const correctness. - Good, I only found one function that could be changed


Testing and Error Checking
1. Unit testing - N/A - there are no unit test files provided within this commit/PR
2. Edge Cases - N/A - there are no unit test files provided within this commit/PR
3. Assertions - There dont seem to be any within these files. 
4. Error handling - Good, errors seem to be handled by returning booleans and printing error messages, or they are handled silently with defaults.

File: source/Interfaces/GUI/interface/Game.cpp (hunk 25 above)
Comment (AbbyFranz) [reply source/Interfaces/GUI/interface/Game.cpp:288]: R286 to R293 also uses this logic

**Comment by MouseCR** [comment]
For Interoperability and API, the demo looks good for showing how the game can be interacted with and what the full game loop looks like. It also is easy to understand how other groups can plug their code into it. I think this satisfies the Interoperability and API part well.

**Comment by MouseCR** [comment]
There is good error handing in Game.cpp but unless i'm just not seeing it I couldn't find a testing file specifically for testing the game class.

**Review by rcm7133 — COMMENTED** [review]
Class design:
- Informative identifiers: Names of variables, functions and types are all clean and easy to understand. If a variable name is not obvious there is plenty of documentation/comments to help
- Avoids Magic Numbers: No magic numbers to be found. Any predefined numbers needed are declared as constexpr, as they should be.
- Intuitive and functional interface: The class is hard to misuse. Good use of private and public functions and member variables. Only what is needed from the user is exposed as public and is hard to misunderstand what each function does. 
- Informative comments: The best use of function documentation and cleanest comments structure for functions and variables I've seen in this project so far. The class is very well organized and broken up with helpful comments to organize functions and variables. Looks great! The comments in game.cpp are much more sparse but the class is designed well enough that comments are not really needed and when they are needed they are used well. Great Job!
- Consistrent style: For the most part this is done well. A couple names like Tiles_x and Tiles_y should be reformatted to be lower case as is done with other local varibles in the same function
- Visual appearance: Extremely readable header file. Very easy to understand and organized very well.  Good use of comment dividers to group functions and variables

Interoperability and API:
- Makes use of stubs and a demo file to show the GUI works with the dungeon world, overworld, and agents. The API is smaller due to the Game class owning pretty much all of the classes. There is not a lot of inter communication for other groups as the Game class is very high level in the project. No real problems with the demo.

File: source/Interfaces/GUI/interface/Game.cpp
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:1]: fixed all - -thank you for you review

Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:31]: added. thanks!

Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:980]: changed

Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:1009]: Good point, this is duplicated logic. 

Though, given time constraints and wanting to avoid introducing regressions before integrating, we kept it inline for now.

but I agree, it would be cleaner as a helper function.

Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp]: the player and goblin now get their own sprites, with dun_monster kept as the fallback for unknown dungeon enemies.

File: source/Interfaces/GUI/interface/Game.cpp (hunk 22 above)
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:968]: renamed.

File: source/Interfaces/GUI/interface/Game.cpp
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:624]: Good point. The update loops are similar, but they currently differ slightly in behavior (like player handling and alive checks)

To avoid introducing bugs this close to integration, I kept them separate, but also agree that this could be refactored into a shared helper in a future pass

File: source/Interfaces/GUI/interface/Game.cpp (hunk 3 above)
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:34]: Agreed. The asset loading section is repetitive and would be cleaner as a table of asset keys/paths loaded in a loop.

Since these keys are used throughout rendering and inventory logic, we are going to leave the current explicit loading in place for stability, but this would be a good future cleanup

File: source/Interfaces/GUI/interface/Game.cpp (hunk 10 above)
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:511]: I addressed part of this, but didn’t fully refactor since it would require larger changes.

File: source/Interfaces/GUI/interface/Game.cpp
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:1039]: maybe a little....
Will leave as is for now due to time constraints :)

File: source/Interfaces/GUI/interface/Game.cpp (hunk 19 above)
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:1029]: switched

File: source/Interfaces/GUI/interface/Game.cpp (hunk 18 above)
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:1014]: Right now I’m just scanning the inventory before and after to detect changes, since it was a simple way to get pickup notifications working

It’s not super efficient, but given the small inventory size it hasn’t been an issue

File: source/Interfaces/GUI/interface/Game.cpp (hunk 12 above)
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:627]: agree

A cleaner setup would be for the world/agent system to expose a “update non-player agents” type helper, but I left it here since that would require changing the world API.

File: source/Interfaces/GUI/interface/Game.cpp (hunk 13 above)
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp]: I hear what you're saying.

a cleaner future version would probably move this into a sprite lookup/helper instead of keeping it inline.

File: source/Interfaces/GUI/interface/Game.cpp (hunk 14 above)
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:717]: The hotbar/backpack rendering grew pretty large as features were added, so if we had time to refactor -- I would make UI file

File: source/Interfaces/GUI/interface/Game.cpp (hunk 9 above)
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:523]: fixed

File: source/Interfaces/GUI/interface/Game.cpp (hunk 11 above)
Comment (Deni-Tepic) [reply source/Interfaces/GUI/interface/Game.cpp:562]: Yeah, it just grew over time. 

Tried to keep the logic grouped by state so it’s at least readable, but breaking input handling into smaller helpers would definitely make this cleaner
